File: 04.Simulation模拟/leetcode_3043_longestCommonPrefix.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int longestCommonPrefix(vector<int>& arr1, vector<int>& arr2) {
        unordered_set<int> st;
        for (int x : arr2){
            for (int y=x; y; y/=10){
                st.insert(y);
            }
        }
        int res = 0;
        for (int x : arr1){
            int length = to_string(x).size();
            for (int y=x; y; y/=10){
                if (st.count(y)) res = max(res, length);
                length--;
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> arr1 = {1,10,100}, arr2 = {1000};
    cout << s.longestCommonPrefix(arr1, arr2) << endl;
}

File: 04.Simulation模拟/leetcode_3043_longestCommonPrefix.py
from typing import List
from collections import defaultdict

class Solution:
    def longestCommonPrefix(self, arr1: List[int], arr2: List[int]) -> int:
        n = len(arr1)
        cnt = defaultdict(int)
        for i in range(n):
            cur = ""
            for j in str(arr1[i]):
                cur += j
                cnt[cur] = 1
        
        res = 0
        for i in range(len(arr2)):
            cur = str(arr2[i])
            for _ in range(len(cur)):
                if cur in cnt:
                    res = max(res, len(cur))
                    break
                cur = cur[:-1]
        return res

if __name__ == "__main__":
    s = Solution()
    arr1 = [1,10,100]
    arr2 = [1000]
    print(s.longestCommonPrefix(arr1, arr2))
